### PR TITLE
OPAL: fix string buffer allocation for large env variables [v4.1.x]

### PR DESCRIPTION
**Pull request by devreal**
In OPAL's keyval parser in `add_to_env_str`, for the first variable a buffer of size 1024 was allocated, regardless of the space needed. For any additional variables, the size might have been doubled at best, even if more space was required. This PR makes sure the allocated buffer is sufficiently large to store the value.

Cherry-pick of https://github.com/open-mpi/ompi/pull/8144 to v4.1.x release branch.

See #8117

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

### CODE REVIEW

**Comment by jsquyres** [comment]
**NOTE:** This is not yet merged on master.  Only merge this PR after #8144 is merged.

**Comment by devreal** [comment]
Updated with latest change from https://github.com/open-mpi/ompi/pull/8144

**Review by jsquyres — APPROVED** [review]

**Pull request merged**
